File: src/main/java/com/kamikazejam/datastore/mode/object/StoreObjectLoader.java
package com.kamikazejam.datastore.mode.object;

import com.google.common.base.Preconditions;
import com.kamikazejam.datastore.base.cache.StoreLoader;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.ref.WeakReference;
import java.util.Optional;

@Getter
@Setter
public class StoreObjectLoader<X extends StoreObject> implements StoreLoader<X> {
    private final StoreObjectCache<X> cache;
    private final String identifier;

    private WeakReference<X> store = null;
    private boolean loadedFromLocal = false;

    StoreObjectLoader(@NotNull StoreObjectCache<X> cache, String identifier) {
        Preconditions.checkNotNull(cache);
        Preconditions.checkNotNull(identifier);
        this.cache = cache;
        this.identifier = identifier;
    }

    @SuppressWarnings("SameParameterValue")
    private void load(boolean fromLocal) {
        if (fromLocal) {
            Optional<X> local = cache.getLocalStore().get(identifier);
            if (local.isPresent()) {
                // Ensure our Store is valid (not recently deleted)
                X store = local.get();
                if (store.isValid()) {
                    this.store = new WeakReference<>(store);
                    loadedFromLocal = true;
                    return;
                }

                // Nullify the reference if the Store is invalid
                // Don't quit, we could in theory still pull from the database
                cache.getLocalStore().remove(identifier);
                this.store = new WeakReference<>(null);
            }
        }

        Optional<X> db = cache.getFromDatabase(identifier, true);
        db.ifPresent(x -> {
            store = new WeakReference<>(x);
            loadedFromLocal = false;
        });
    }

    @Override
    public Optional<X> fetch(boolean saveToLocalCache) {
        load(true);

        if (store == null) {
            return Optional.empty();
        }

        // Double check validity here too
        @Nullable X p = store.get();
        if (p != null && !p.isValid()) {
            store = new WeakReference<>(null);
            return Optional.empty();
        }

        // Save to local cache if necessary
        if (saveToLocalCache && p != null && !loadedFromLocal) {
            this.cache.cache(p);
        }

        // Ensure the Store has its cache set
        if (p != null) {
            p.setCache(cache);
        }
        return Optional.ofNullable(p);
    }
}

File: src/main/kotlin/com/kamikazejam/datastore/mode/object/StoreObjectLoader.kt
package com.kamikazejam.datastore.mode.`object`

import com.google.common.base.Preconditions
import com.kamikazejam.datastore.base.loader.StoreLoader
import com.kamikazejam.datastore.mode.store.StoreObject
import java.lang.ref.WeakReference

class StoreObjectLoader<X : StoreObject<X>> internal constructor(collection: StoreObjectCollection<X>, identifier: String) : StoreLoader<X> {
    private val collection: StoreObjectCollection<X>
    private val identifier: String

    private var store: WeakReference<X>? = null
    private var loadedFromLocal = false

    init {
        Preconditions.checkNotNull(collection)
        Preconditions.checkNotNull(identifier)
        this.collection = collection
        this.identifier = identifier
    }

    @Suppress("SameParameterValue")
    private suspend fun load(fromLocal: Boolean) {
        if (fromLocal) {
            val local: X? = collection.localStore.get(identifier)
            if (local != null) {
                // Ensure our Store is valid (not recently deleted)
                if (local.valid) {
                    this.store = WeakReference(local)
                    loadedFromLocal = true
                    return
                }

                // Nullify the reference if the Store is invalid
                // Don't quit, we could in theory still pull from the database
                collection.localStore.remove(identifier)
                this.store = WeakReference(null)
            }
        }

        val db = collection.readFromDatabase(identifier, true)
        db?.let { x ->
            store = WeakReference(x)
            loadedFromLocal = false
        }
    }

    override suspend fun fetch(saveToLocalCache: Boolean): X? {
        load(true)

        // Double check validity here too
        store?.let { ref ->
            val p = ref.get()
            if (p != null && !p.valid) {
                store = WeakReference(null)
                return null
            }

            // Save to local if necessary
            if (saveToLocalCache && p != null && !loadedFromLocal) {
                collection.cache(p)
            }

            // Ensure the Store has its collection set
            p?.initialize(collection)
            return p
        }

        return null
    }
}
